### PR TITLE
Configure workload identity

### PR DESCRIPTION
**Pull request by SulakshanaBandara**
Description:

- Update the AKS (Azure Kubernetes Service) cluster and configuring federated identities. The goal is to integrate Azure AD Workload Identity.

Changes made:

- Integrate a new resource as Federated_identity_credentials.
- Updated the AKS cluster by enabling workload identity
- Updated the AKS cluster by enabling OIDC url in the cluster.

Related GIT issue:
[https://github.com/wso2-enterprise/iam-engineering/issues/863](https://github.com/wso2-enterprise/iam-engineering/issues/863)

### CODE REVIEW

File: modules/azurerm/AKS-Firewall/aks_cluster.tf [hunk 1]
```diff
@@ -26,6 +26,8 @@ resource "azurerm_kubernetes_cluster" "aks_cluster" {
   http_application_routing_enabled    = var.http_application_routing_enabled
   tags                                = var.tags
   depends_on                          = [azurerm_subnet.aks_node_pool_subnet]
+  workload_identity_enabled           = true
```
Comment (Uvindu96) [inline modules/azurerm/AKS-Firewall/aks_cluster.tf:29]: better to parameterize these values

File: modules/azurerm/Federated-Identity/Federated-Credentials.tf [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+# --------------------------------------------------------------------------------------
+
+resource "azurerm_federated_identity_credential" "user_assigned_identity" {
+  name                = var.federated_identity_name
+  resource_group_name = var.resource_group_name
+  parent_id           = var.azurerm_user_assigned_identity_principal_id
+  audience            = ["api://AzureADTokenExchange"]
+  issuer              = var.oidc_issuer_url
+  subject             = "system:serviceaccount:${workload_sa_namespace}:${workload_sa_name}"
+}
```
Comment (Uvindu96) [inline modules/azurerm/Federated-Identity/Federated-Credentials.tf:19]: need a new line here

File: modules/azurerm/AKS-Firewall/variables.tf [hunk 3]
```diff
@@ -284,3 +284,13 @@ variable "nsg_rule_apgw_inbound_allow_enabled" {
...
+
+variable "workload_identity_enabled" {
+  description = "Enable workload identity in AKS cluster"
+  type        = bool
+}
+
+variable "oidc_issuer_enabled" {
+  description = "Enable OIDC URL in AKS cluster"
+  type        = bool
+}
```
Comment (ayeshajay) [inline modules/azurerm/AKS-Firewall/variables.tf:296]: Add a new line

File: modules/azurerm/Federated-Identity/Federated-Credentials.tf [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
```
Comment (ayeshajay) [inline modules/azurerm/Federated-Identity/Federated-Credentials.tf:20]: remove unwanted new line

File: modules/azurerm/Key-Vault-Access-Policy/variables.tf [hunk 5]
```diff
@@ -48,3 +48,5 @@ variable "key_vault_id" {
   description = "value of the key_vault_id"
   type        = string
 }
+
+variable "" {}
```
Comment (ayeshajay) [inline modules/azurerm/Key-Vault-Access-Policy/variables.tf:52]: Do we need this?

**Review by ayeshajay — CHANGES_REQUESTED** [review]

File: modules/azurerm/AKS-Firewall/variables.tf [hunk 6]
```diff
@@ -284,3 +284,13 @@ variable "nsg_rule_apgw_inbound_allow_enabled" {
   type        = bool
   default     = false
 }
+
+variable "workload_identity_enabled" {
+  description = "Enable workload identity in AKS cluster"
+  type        = bool
```
Comment (Migara-Pramod) [inline modules/azurerm/AKS-Firewall/variables.tf:290]: Set a default value to false.

File: modules/azurerm/AKS-Firewall/variables.tf [hunk 7]
```diff
@@ -284,3 +284,13 @@ variable "nsg_rule_apgw_inbound_allow_enabled" {
...
+}
+
+variable "oidc_issuer_enabled" {
+  description = "Enable OIDC URL in AKS cluster"
+  type        = bool
```
Comment (Migara-Pramod) [inline modules/azurerm/AKS-Firewall/variables.tf:295]: Set default value to false

File: modules/azurerm/Federated-Identity/Federated-Credentials.tf [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+# Copyright (c) 2024, WSO2 LLC. (http://www.wso2.com). All Rights Reserved.
+#
+# This software is the property of WSO2 LLC. and its suppliers, if any.
+# Dissemination of any information or reproduction of any material contained
+# herein in any form is strictly forbidden, unless permitted by WSO2 expressly.
+# You may not alter or remove any copyright or other notice from copies of this content.
+#
+# --------------------------------------------------------------------------------------
+
+resource "azurerm_federated_identity_credential" "federated_credentials_name" {
```
Comment (Migara-Pramod) [inline modules/azurerm/Federated-Identity/Federated-Credentials.tf:12]:
```suggestion
resource "azurerm_federated_identity_credential" "federated_credentials" {
```

File: modules/azurerm/Federated-Identity/Federated-Credentials.tf [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  name                = var.federated_identity_name
+  resource_group_name = var.resource_group_name
+  parent_id           = var.azurerm_user_assigned_identity_principal_id
+  audience            = ["api://AzureADTokenExchange"]
```
Comment (Migara-Pramod) [inline modules/azurerm/Federated-Identity/Federated-Credentials.tf:16]: Why we hardcode the audience

File: modules/azurerm/Federated-Identity/Federated-Credentials.tf [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  issuer              = var.oidc_issuer_url
+  subject             = "system:serviceaccount:${workload_sa_namespace}:${workload_sa_name}"
```
Comment (Migara-Pramod) [inline modules/azurerm/Federated-Identity/Federated-Credentials.tf:18]: Shall we get this from a variable as well

File: modules/azurerm/Federated-Identity/variables.tf [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  type        = string
+}
+
+variable "aks_cluster_name" {
+  description = "AKS cluster name"
+  type        = string
+}
+
+variable "federated_identity_name" {
+  description = "AKS cluster name"
```
Comment (Migara-Pramod) [inline modules/azurerm/Federated-Identity/variables.tf:23]: Change this

File: modules/azurerm/Federated-Identity/variables.tf [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+  type        = string
+}
+
+variable "azurerm_user_assigned_identity_principal_id" {
+  description = "User assigned managed identity client ID"
+  type        = string
+}
+
+variable "oidc_issuer_url" {
+  description = "OIDC Url"
```
Comment (Migara-Pramod) [inline modules/azurerm/Federated-Identity/variables.tf:33]:
```suggestion
  description = "OIDC issuer url"
```

**Review by Migara-Pramod — CHANGES_REQUESTED** [review]

**Review by Migara-Pramod — CHANGES_REQUESTED** [review]

**Pull request closed without merging**